lib/argonaut/reservations.ex

```diff
@@ -0,0 +1,241 @@
+        Repo.delete(reservation)
+      end
+
+      new_reservation = Repo.insert!(%Reservation{
+        user_id: user.id,
+        environment_id: environment.id,
+        application_id: application.id,
+        team_id: environment.team_id,
+        reserved_at: DateTime.utc_now()
+      })
+
+      reservation_tree = reservation_with_associations(new_reservation.id)
+      ArgonautWeb.Endpoint.broadcast!("teams:#{environment.team_id}", "reservation_created", reservation_tree)
+
+      true
+    else
+      false
+    end
+  end
+
+  defp can_reserve?(nil, _), do: true
+  defp can_reserve?(_, _), do: false
+
+  defp can_release?(nil, _), do: false
+  defp can_release?(%Reservation{user_id: user_id}, %User{id: id}), do: user_id == id
+
+  def delete_reservation(%{"app" => app, "env" => env}, current_user) do
+    environment = Repo.one(from(env in Environment, where: env.name == ^env))
+
+    application =
+      Repo.one(
+        from(
+          app in Application,
+          where: app.name == ^app,
+          where: app.team_id == ^environment.team_id
+        )
+      )
+
+    reservation =
+      from(
+        r in Reservation,
+        where: r.application_id == ^application.id,
+        where: r.environment_id == ^environment.id
+      )
+      |> Repo.one()
+
+    if can_release?(reservation, current_user) do
+      Repo.delete(reservation)
+
+      reservation_tree = %{reservation | application: application, environment: environment, user: current_user}
+      ArgonautWeb.Endpoint.broadcast!("teams:#{environment.team_id}", "reservation_deleted", reservation_tree)
+
+      %{message: "Deleted your reservation on #{env}:#{app}", success: true}
+    else
+      %{message: "Could not delete reservation on #{env}:#{app}", success: false}
+    end
+  end
+
+  def find_application(%{"app" => app}) do
+    applications =
+      from(
+        a in Application,
+        where: a.name == ^app,
+        join: r in Reservation,
+        on: a.id == r.application_id,
+        join: e in Environment,
+        on: e.id == r.environment_id,
+        preload: [reservations: a]
+      )
+      |> Repo.all()
+
+    applications
+    |> Enum.map(fn x -> %{application: x.application.name, environment: x.environment.name} end)
+  end
+
+  def clear_user_reservations(current_user) do
+    {deleted_count, _} =
+      from(r in Reservation, where: r.user_id == ^current_user.id) |> Repo.delete_all()
+
+    %{success: true, message: "Cleared all (#{deleted_count}) reservations"}
+  end
+
+  def list_user_reservations(current_user) do
+    reservations =
+      reservations_by_user(current_user.id)
+      |> Repo.all()
+      |> Enum.map(fn r ->
+        %{
+          environment: r.environment.name,
+          application: r.application.name,
+          reserved_at: r.reserved_at
+        }
+      end)
+
+    %{
+      success: true,
+      message: "List of reservations made by you (#{current_user.username})",
+      data: reservations
+    }
```

Comment:
Not sure where you're using this, but it looks like you're formatting your data for output here. I'd do that in the controller type thing, or better yet, a view.